### PR TITLE
Check for null + empty strings

### PR DESCRIPTION
**Pull request by m0wfo**

### CODE REVIEW

File: src/main/java/jetcd/EtcdClientImpl.java [hunk 1]
```diff
@@ -45,8 +46,8 @@ public String get(String key) throws EtcdException {
 
     @Override
     public void set(String key, String value) throws EtcdException {
-        Preconditions.checkNotNull(key);
-        Preconditions.checkNotNull(value);
+		Preconditions.checkArgument(!Strings.isNullOrEmpty(key));
```
Comment (diwakergupta) [inline src/main/java/jetcd/EtcdClientImpl.java]: bad indent (or tab?)

File: src/main/java/jetcd/EtcdClientImpl.java [hunk 2]
```diff
@@ -56,8 +57,8 @@ public void set(String key, String value) throws EtcdException {
 
     @Override
     public void set(String key, String value, int ttl) throws EtcdException {
-        Preconditions.checkNotNull(key);
-        Preconditions.checkNotNull(value);
+		Preconditions.checkArgument(!Strings.isNullOrEmpty(key));
```
Comment (diwakergupta) [inline src/main/java/jetcd/EtcdClientImpl.java]: Fix indent

**Comment by diwakergupta** [comment]
jetcd's first contribution -- thanks!! LG modulo indentation. Also, Guava offers this tip when using isEmptyOrNull:

> Consider normalizing your string references with nullToEmpty(java.lang.String). If you do, you can use String.isEmpty() instead of this method, and you won't need special null-safe forms of methods like String.toUpperCase(java.util.Locale) either. Or, if you'd like to normalize "in the other direction," converting empty strings to null, you can use emptyToNull(java.lang.String).

Thoughts?

File: src/main/java/jetcd/EtcdClientImpl.java (hunk 1 above)
Comment (m0wfo) [inline src/main/java/jetcd/EtcdClientImpl.java]: Just moved to IDEA, which seems to default to 4-space indentation; not sure what's messing the formatting up. I'll have a play around :/

Comment (m0wfo) [inline src/main/java/jetcd/EtcdClientImpl.java]: turns out I'm half asleep, c.f. ad40646

**Comment by diwakergupta** [comment]
@m0wfo thanks and sorry for the delay -- github somehow ate some of the updates! Can you update the description of the pull request? If you don't mind pushing separate PRs for formatting changes and functional changes, that would be ideal.

I really like Typesafe's config as well (I introduced it at work) but I'm not sure exactly what you are trying to achieve. You want to configure jetcd client using a Config object as input?

File: src/main/java/jetcd/EtcdClient.java [hunk 3]
```diff
@@ -64,6 +64,19 @@
...
+	 *
+	 * <p>Unlike {@link #list(String)}, this recursively iterates
+	 * through directory children, building nested maps for any
+	 * sub-directories it finds.</p>
+	 *
+	 * @param path Given path
+	 * @return Map of key,values pairs under the path
+	 * @throws EtcdException
+	 */
+	Map<String, Object> deepList(String path) throws EtcdException;
```
Comment (diwakergupta) [inline src/main/java/jetcd/EtcdClient.java:78]: For EtcdClient I'd like to stay as close to the API exposed by etcd directly (I refer to go-etcd for reference). This functionality seems useful so maybe add it in a separate helper/util class? Another good analog is ZooKeeper and higher-order primitives built on top of ZK like Curator. WDYT?

**Comment by m0wfo** [comment]
@diwakergupta I'm trying to ultimately replace a json/xml file with config pulled from an etcd node. For that to happen, I need a fully populated map (i.e. with all child paths/values) which I pass into a [ConfigValueFactory](http://typesafehub.github.io/config/latest/api/com/typesafe/config/ConfigValueFactory.html). Unless there's a simpler way to do it.

You're right, I guess this kind of behaviour should be separated from the low-level api. I think ZK and Curator is a good aspirational comparison.

I'll close & submit separate PRs when I get time over the next couple of days.

**Comment by diwakergupta** [comment]
@m0wfo ah thanks. Instead of storing individual entries under separate paths, you can always just store a serialized JSON blob. Any particular reason why you want to store the expanded config? I guess one pro is that you can subscribe to changes at a much finer granularity.

BTW are you actually using jetcd? Would love to hear about your use case :-) I plan to push some more enhancements this weekend, stay tuned!

**Comment by m0wfo** [comment]
We're [Logentries] using the Config library for a couple of services in production already and are playing around with deploying artifacts as docker images on coreos / ubuntu. Purely experimental right now, but I'll keep you in the loop.

**Pull request closed without merging**